doc/design/graph.md

```diff
@@ -8,6 +8,8 @@ This document explains that the construction of a graph as three steps:
 - construct the backward part
 - construct the optimization part
 
+## The Construction of a Graph
```

Comment:
I think the training needs 1) the cost, and 2) the parameter to be optimized to minimize the cost.

The cost is specified in the invocation to `train`.

Parameters could be created by a layer function like `layer.fc`, or the user via `W = paddle.Var(type=parameter, ...)`. Anyway, they are marked parameters and can be updated.

Reply:
> the training needs 1) the cost, and 2) the parameter to be optimized to minimize the cost.

I think it need the optimizer as well (Adam or Adagrad).
For example, it user do something like:
```Python
opt0 = pd.Adam(cost)
opt1 = pd.Adagrad(cost)
train(cost, reader=mnist.train())
```
What optimizer will Paddle use for training?
However, maybe the Python code you wrote is the V2 API, which maybe only allow one optimizer to be connected with the cost.

Reply:
Yes. What I mean is that we can have two forms of `Block::Eval`:

1. One accepts targets of type `Variable`s: 

   ```c++
   void Block::Eval(vector<Variable*> targets);
   ```

   which is used to do forward computation.  It traces only operators in `BlockDesc::ops` before `targets`.

    1. *Forward computation*: Because our Python API doesn't expose gradient variables to users, `targets` have to be forward variables, so this form of `Block::Eval` works only with forward computation.

   1. *Backward computation*: In the C++ world, `Block::Eval` can accept gradient variables as its targets. We can create a Python API function, say `backward`, which calls `Block::Eval` with gradient variables to do the backward computation.

1. The other form of `Block::Eval` accepts targets as operators:

   ```c++
   void Block::Eval(vector<Operator*> targets);
   ```

   Somewhere in the C++ world, we can enumerate all optimization operators and use them as the target, so could we run the optimization step.